### PR TITLE
common: kill the bl::last_p member. Use explicit hint passing instead.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>

Potential beneficiaries of the `last_p` - to check
--------------------------------------------------------------

 - [x] `src/client/Client.cc:13087`
 - [x] `src/client/Client.cc:13308`
 - [x] `src/client/Client.cc:13315`
 - [x] `src/client/Client.cc:8927`
 - [x] `src/client/Client.cc:9297`
 - [x] `src/client/Client.cc:9300`
 - [x] `src/client/Client.cc:9438`
 - [x] `src/common/bloom_filter.cc:41`
 - [ ] `src/common/buffer.cc:1551`
 - [x] `src/crush/CrushLocation.cc:92`
 - [ ] `src/erasure-code/ErasureCode.cc:157`
 - [x] `src/libcephfs.cc:1570`
 - [x] `src/librados/librados.cc:3874`
 - [x] `src/librados/librados.cc:3906`
 - [x] `src/librados/librados.cc:4319`
 - [x] `src/librados/librados.cc:4468`
 - [x] `src/librados/librados.cc:4500`
 - [x] `src/librados/librados.cc:5035`
 - [x] `src/librados/librados.cc:6099`
 - [x] `src/libradosstriper/libradosstriper.cc:435`
 - [x] `src/libradosstriper/libradosstriper.cc:466`
 - [x] `src/mds/Server.cc:5191`
 - [x] `src/mgr/ActivePyModule.cc:182`
 - [x] `src/os/bluestore/BlueStore.cc:10351`
 - [ ] `src/os/bluestore/BlueStore.cc:10374`
 - [ ] `src/os/bluestore/NVMEDevice.cc:539`
 - [ ] `src/os/bluestore/NVMEDevice.cc:571`
 - [x] `src/osdc/Objecter.h:3039`
 - [x] `src/osdc/Striper.cc:400`
 - [x] `src/osdc/Striper.cc:403`
 - [x] `src/osd/osd_types.cc:6158`
 - [x] `src/osd/osd_types.cc:6168`
 - [ ] `src/osd/PrimaryLogPG.cc:3095`
 - [x] `src/os/ObjectStore.h:747`
 - [x] `src/rgw/rgw_compression.cc:102`
 - [ ] `src/rgw/rgw_compression.cc:80`
 - [x] `src/rgw/rgw_compression.cc:98`
 - [ ] `src/rgw/rgw_crypt.cc:617`
 - [x] `src/rgw/rgw_op.cc:3434`
 - [x] `src/rgw/rgw_rados.cc:10589`

Potential beneficiaries of the `last_p` - to converge
------------------
 - [ ] `src/client/Client.cc:9297`
 - [ ] `src/client/Client.cc:9300`
 - [x] `src/osdc/Objecter.h:3039`
 - [ ] `src/rgw/rgw_compression.cc:102`
 - [ ] `src/rgw/rgw_compression.cc:98`

### CODE REVIEW

File: src/include/buffer.h [hunk 1]
```diff
@@ -874,6 +869,22 @@ namespace buffer CEPH_BUFFER_API {
...
     void copy_in(unsigned off, unsigned len, const list& src);
 
+    typedef iterator iter_hint_t;
+    typedef const_iterator constiter_hint_t;
+
+    iter_hint_t create_iter_hint(unsigned off = 0) {
+      return iter_hint_t(this, 0);
+    }
+    constiter_hint_t create_iter_hint(unsigned off = 0) const {
+      return constiter_hint_t(this, 0);
```
Comment (cbodley) [inline src/include/buffer.h]: `(this, off)` ?

Comment (rzarzynski) [reply src/include/buffer.h]: Definitely! Thanks.

**Comment by gregsfortytwo** [comment]
Hmm, the existence of ```last_p``` seems like an awfully micro-optimization. What motivates you to kill it?

Without any users of this iterator hint infrastructure, it seems pretty pointless to have anything like it, instead of just freshly-calculating the buffer offsets each time...

**Comment by rzarzynski** [comment]
As @cbodley points out, the idea is to pay for something only when we actually use it. Compiler has testified  all variants of `bufferlist::copy_in` and `bufferlist::copy` have no more than 40 users (vanilla `./do_cmake.sh`). Manual inspection shows only very few of them pass non-`0` as the `off` argument. Refactoring in such situation looks appealing. The changeset switched `src/osdc/Objecter.h:3039` as an example of hint usage.

Part of the price is being paid on `bufferlist` construction. All our constructors, including the moving one, does `last_p(this)`. This basically leads to `::advance(0)` but the path is surprisingly long:

```
0000000000c90400 <ceph::buffer::list::list(ceph::buffer::list&&)>:
  ...
  c90484:       31 d2                   xor    %edx,%edx
  c90486:       e8 e5 b1 ff ff          callq  c8b670 <ceph::buffer::list::iterator::iterator(ceph::buffer::list*, unsigned int)>
 
 
0000000000c8b670 <ceph::buffer::list::iterator::iterator(ceph::buffer::list*, unsigned int)>:
  c8b670:       e9 0b 81 00 00          jmpq   c93780 <ceph::buffer::list::iterator_impl<false>::iterator_impl(ceph::buffer::list*, unsigned int)>
  c8b675:       90                      nop
  c8b676:       66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
  c8b67d:       00 00 00 

0000000000c93780 <ceph::buffer::list::iterator_impl<false>::iterator_impl(ceph::buffer::list*, unsigned int)>:
  c93780:       48 89 37                mov    %rsi,(%rdi)
  c93783:       48 89 77 08             mov    %rsi,0x8(%rdi)
  c93787:       c7 47 10 00 00 00 00    movl   $0x0,0x10(%rdi)
  c9378e:       48 8b 06                mov    (%rsi),%rax
  c93791:       89 d6                   mov    %edx,%esi
  c93793:       c7 47 20 00 00 00 00    movl   $0x0,0x20(%rdi)
  c9379a:       48 89 47 18             mov    %rax,0x18(%rdi)
  c9379e:       e9 0d ff ff ff          jmpq   c936b0 <ceph::buffer::list::iterator_impl<false>::advance(int)>
  c937a3:       66 2e 0f 1f 84 00 00    nopw   %cs:0x0(%rax,%rax,1)
  c937aa:       00 00 00 
  c937ad:       0f 1f 00                nopl   (%rax)

0000000000c936b0 <ceph::buffer::list::iterator_impl<false>::advance(int)>:
  c936b0:       48 83 ec 08             sub    $0x8,%rsp
  c936b4:       85 f6                   test   %esi,%esi
  c936b6:       8b 47 20                mov    0x20(%rdi),%eax
  c936b9:       78 35                   js     c936f0 <ceph::buffer::list::iterator_impl<false>::advance(int)+0x40>
  c936bb:       4c 8b 47 08             mov    0x8(%rdi),%r8
  c936bf:       48 8b 57 18             mov    0x18(%rdi),%rdx
  c936c3:       01 f0                   add    %esi,%eax
  c936c5:       89 47 20                mov    %eax,0x20(%rdi)
  c936c8:       4c 39 c2                cmp    %r8,%rdx
  c936cb:       75 14                   jne    c936e1 <ceph::buffer::list::iterator_impl<false>::advance(int)+0x31>
  c936cd:       eb 71                   jmp    c93740 <ceph::buffer::list::iterator_impl<false>::advance(int)+0x90>
  c936cf:       90                      nop
  ...
  c936e8:       01 77 10                add    %esi,0x10(%rdi)
  c936eb:       48 83 c4 08             add    $0x8,%rsp
  c936ef:       c3                      retq   
  ...
  c93740:       85 c0                   test   %eax,%eax
  c93742:       74 a4                   je     c936e8 <ceph::buffer::list::iterator_impl<false>::advance(int)+0x38>
```

The delegate jumping might be a bit tricky for profiler as it avoids stack frame.

At the moment I don't have numbers, so the PR isn't labelled as *performance*. However, even if there is absolutely no difference, I still would love to drop `last_p` for the sake of cleaning the stuff up.

> i assumed that @rzarzynski was compiling an inventory of callers that rely on this, in order to include the hints as part of this pr. is that right?

Right. In case of doubts/harder cases the list might be useful when calling for component lead's opinion, I guess.

**Comment by gregsfortytwo** [comment]
I presume this also breaks the API/ABI just like everything else we want to do, though?
Otherwise I have no particular thoughts on this.

**Comment by rzarzynski** [comment]
@tchaikov: this one is pretty small and likely harmless. I would love to see it in nautilus.
There are also two bigger things ready: https://github.com/ceph/ceph/pull/25077 and https://github.com/ceph/ceph/pull/25082.

**Comment by rzarzynski** [comment]
I think we can close this PR as https://github.com/ceph/ceph/pull/32831 is merged.

**Pull request closed without merging**